consumer/pdatautil/metricsbuilder.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2020 The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pdatautil
+
+import (
+	"time"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/internal/data"
+)
+
+// MetricsBuilder facilitates building pdata.Metrics in receivers.  It is meant
+// to be much easier and more fluent than than using pdata.Metrics directly.
+// All of the methods on it except for Flush() will return the same instance of
+// MetricsBuilder as a return value, allowing you to chain method calls easily,
+// similar to the Java builder pattern.
+//
+// The MetricsBuilder is designed for cases where receivers are generating
+// metrics from scratch, where generally you will have a single datapoint per
+// metric/label combination.
+//
+// It is NOT thread-safe, so you should use an external mutex if using it from
+// multiple goroutines.
+type MetricsBuilder struct {
+	// If set, this instrumentation library name will be used for all metrics
+	// generated by this builder.
+	InstrumentationLibraryName string
+	// If set, this instrumentation library version will be used for all metrics
+	// generated by this builder.
+	InstrumentationLibraryVersion string
+	// These attributes will be added to the Resource object on all
+	// ResourceMetrics instances created by the builder.
+	ResourceAttributes map[string]string
+	// This time will be used as the Timestamp for all metrics generated.  It
+	// can be updated with a new timestamp at any time.
+	Timestamp time.Time
+
+	metrics pdata.MetricSlice
+	labels  map[string]string
+	isInit  bool
+}
+
+// Flush creates a pdata.Metrics instance with all of the pending metrics on
+// the builder and returns it.  Subsequent calls to Flush will only return
+// metrics added after the previous Flush call.  Labels set with With*Labels
+// will be retained after a call to Flush.
+func (mb *MetricsBuilder) Flush() pdata.Metrics {
+	md := data.NewMetricData()
+	rms := md.ResourceMetrics()
+	rms.Resize(1)
+	rm := rms.At(0)
+	rm.InitEmpty()
+
+	res := rm.Resource()
+	res.InitEmpty()
+	for k, v := range mb.ResourceAttributes {
+		res.Attributes().Insert(k, pdata.NewAttributeValueString(v))
```

Comment:
Can we initialize the Attributes() map using InitEmptyWithCapacity() before the for loop?